File: android/src/main/java/com/dahabtech/mediaeditor/PhotoFileWriter.kt
package com.dahabtech.mediaeditor

import android.content.Context
import android.net.Uri
import android.util.Base64
import java.io.File
import java.io.FileOutputStream

object PhotoFileWriter {
  fun write(context: Context, base64: String, extension: String): String {
    val bytes = try {
      Base64.decode(base64, Base64.DEFAULT)
    } catch (exception: IllegalArgumentException) {
      throw MediaEditorException("ERR_WRITE_FILE", "Invalid base64 payload")
    }
    val outputFile = cacheFile(context, "photo", extension)
    FileOutputStream(outputFile).use { stream ->
      stream.write(bytes)
    }
    return Uri.fromFile(outputFile).toString()
  }

  /** Read a `file://` URI (or bare path) and return the base64-encoded contents. */
  fun readAsBase64(uri: String): String {
    val parsed = Uri.parse(uri)
    val path = when (parsed.scheme) {
      null -> uri
      "file" -> parsed.path
      else -> null
    } ?: throw MediaEditorException("ERR_READ_FILE", "Unsupported URI scheme: $uri")
    val file = File(path)
    if (!file.exists()) {
      throw MediaEditorException("ERR_READ_FILE", "No file exists at $uri")
    }
    // NO_WRAP matches the shape writeCacheFile consumers (Skia's encodeToBase64) produce.
    return Base64.encodeToString(file.readBytes(), Base64.NO_WRAP)
  }
}

File: ios/PhotoFileWriter.swift
import Foundation

struct PhotoFileWriter {
  static func write(base64: String, fileExtension: String) throws -> URL {
    guard let data = Data(base64Encoded: base64, options: .ignoreUnknownCharacters) else {
      throw MediaEditorError.exportFailed("Invalid base64 payload")
    }
    let outputURL = try CacheFile.make(prefix: "photo", fileExtension: fileExtension)
    try data.write(to: outputURL, options: .atomic)
    return outputURL
  }

  /// Read a `file://` URI (or bare path) and return the base64-encoded contents.
  static func readAsBase64(uri: String) throws -> String {
    let url: URL
    if let parsed = URL(string: uri), parsed.scheme != nil {
      url = parsed
    } else {
      url = URL(fileURLWithPath: uri)
    }
    let data = try Data(contentsOf: url)
    return data.base64EncodedString()
  }
}
